aptos-move/framework/src/natives/cryptography/algebra/mod.rs

```diff
@@ -0,0 +1,319 @@
+// Copyright © Aptos Foundation
+
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    natives::{
+        cryptography::algebra::gas::GasParameters,
+        helpers::{make_safe_native, SafeNativeContext, SafeNativeError, SafeNativeResult},
+    },
+    safely_pop_arg,
+};
+use aptos_types::on_chain_config::{FeatureFlag, Features, TimedFeatures};
+use ark_serialize::{CanonicalDeserialize, CanonicalSerialize};
+use better_any::{Tid, TidAble};
+use move_core_types::language_storage::TypeTag;
+use move_vm_runtime::native_functions::NativeFunction;
+use move_vm_types::{
+    loaded_data::runtime_types::Type,
+    values::{Value, VectorRef},
+};
+use smallvec::{smallvec, SmallVec};
+use std::{any::Any, collections::VecDeque, hash::Hash, ops::Add, rc::Rc, sync::Arc};
+
+pub mod gas;
+
+/// Equivalent to `std::error::not_implemented(0)` in Move.
+const MOVE_ABORT_CODE_NOT_IMPLEMENTED: u64 = 0x0C0000;
+
+#[derive(Copy, Clone, Eq, Hash, PartialEq)]
+pub enum Structure {
+    BLS12381Fq12,
+    BLS12381G1Affine,
+    BLS12381G2Affine,
+    BLS12381Gt,
+    BLS12381Fr,
+}
+
+impl Structure {
+    pub fn from_type_tag(type_tag: &TypeTag) -> Option<Structure> {
+        match type_tag.to_string().as_str() {
+            "0x1::algebra_bls12381::Fr" => Some(Structure::BLS12381Fr),
+            "0x1::algebra_bls12381::Fq12" => Some(Structure::BLS12381Fq12),
+            "0x1::algebra_bls12381::G1Affine" => Some(Structure::BLS12381G1Affine),
+            "0x1::algebra_bls12381::G2Affine" => Some(Structure::BLS12381G2Affine),
+            "0x1::algebra_bls12381::Gt" => Some(Structure::BLS12381Gt),
+            _ => None,
+        }
+    }
+}
+
+#[derive(Copy, Clone, Eq, Hash, PartialEq)]
+pub enum SerializationFormat {
+    BLS12381Fq12LscLscLscLsb,
+    BLS12381G1AffineCompressed,
+    BLS12381G1AffineUncompressed,
+    BLS12381G2AffineCompressed,
+    BLS12381G2AffineUnompressed,
+    BLS12381Gt,
+    BLS12381FrLsb,
+    BLS12381FrMsb,
+}
+
+impl TryFrom<u64> for SerializationFormat {
+    type Error = ();
+
+    fn try_from(value: u64) -> Result<Self, Self::Error> {
+        match value {
```

Comment:
Can you explain where all these constants came from & what do they mean?

Reply:
Fixed.